### PR TITLE
Refactor solution views

### PR DESCRIPTION
**Pull request by PhilippMatthes**
Closes: #272

### CODE REVIEW

**Comment by PhilippMatthes** [comment]
@martinmo in the course of reviewing, you can pull the nested error classes out, by pushing directly onto this branch, if you want to.

File: inloop/solutions/views.py [hunk 1]
```diff
@@ -29,54 +29,111 @@ def get(self, request, id):
         return JsonResponse({'solution_id': solution.id, 'status': solution.status()})
 
 
-class SolutionEditorView(LoginRequiredMixin, View):
+class SolutionSubmissionView(LoginRequiredMixin, View):
```
Comment (martinmo) [inline inloop/solutions/views.py]: I think the name is misleading, this class is not a view, but just a base class with some methods that are used in views. Probably should also end in `Mixin`.

File: inloop/solutions/views.py [hunk 2]
```diff
@@ -29,54 +29,111 @@ def get(self, request, id):
...
-            return failure_response
+            return SolutionEditorView.JsonFailureResponse()
 
-        if task.is_expired:
-            messages.error(request, 'The deadline for this task has passed.')
-            return failure_response
+        try:
+            task = self.get_task(slug)
+        except SolutionSubmissionView.Error as e:
+            messages.error(request, e)
```
Comment (martinmo) [inline inloop/solutions/views.py]: Messages don't make sense in this context – there are used in HTML views only and are displayed after a redirect.

File: inloop/solutions/views.py [hunk 3]
```diff
@@ -29,54 +29,111 @@ def get(self, request, id):
...
+            super().__init__({'success': True})
+
     def get(self, request, slug):
         return TemplateResponse(request, 'solutions/editor/editor.html', {
             'task': get_object_or_404(Task.objects.published(), slug=slug),
             'active_tab': 1
         })
 
     def post(self, request, slug):
-        task = get_object_or_404(Task.objects.published(), slug=slug)
```
Comment (martinmo) [inline inloop/solutions/views.py]: Behavior changed from returning a `404` to returning a `200` with a JSON error message (so technically, this isn't a refactoring).

File: inloop/solutions/views.py [hunk 4]
```diff
@@ -135,47 +192,6 @@ def get(self, request, slug):
...
 
-class SolutionUploadView(LoginRequiredMixin, View):
-    def get(self, request, slug):
-        return TemplateResponse(request, 'solutions/upload_form.html', {
-            'task': get_object_or_404(Task.objects.published(), slug=slug),
-            'active_tab': 2
-        })
-
-    def post(self, request, slug):
-        task = get_object_or_404(Task.objects.published(), slug=slug)
```
Comment (martinmo) [inline inloop/solutions/views.py]: Same here, behavior changed from returning a `404` to returning a `302`

**Pull request merged**